storage/src/linear/io_uring.rs

```diff
@@ -0,0 +1,811 @@
+                firewood_counter!(
+                    "ring.eagin_write_retry",
+                    "amount of io-uring write entries that have been re-submitted due to EAGIN io error"
+                ).increment(1);
+                return Ok(0);
+            }
+
+            return Err(BatchError {
+                batch_offset: self.original_offset,
+                error_offset: self.offset,
+                err,
+                incurable: false,
+            });
+        }
+
+        // This check is to prevent infinite loops. The kernel should never
+        // return Ok(0) for a write operation via this API unless the storage
+        // driver is broken (or we weren't writing to a file--how did we get here?)
+        if res == 0 {
+            return Err(BatchError {
+                batch_offset: self.original_offset,
+                error_offset: self.offset,
+                err: std::io::Error::new(std::io::ErrorKind::WriteZero, "kernel wrote zero bytes"),
+                incurable: false,
+            });
+        }
+
+        let written = res as u32;
+        if written > self.length {
+            return Err(BatchError {
+                batch_offset: self.original_offset,
+                error_offset: self.offset,
+                err: std::io::Error::other("kernel wrote more data than requested"),
+                incurable: false,
+            });
+        }
+
+        // if this is non-zero, we need to re-submit the remaining data
+        self.length -= written;
+        self.pointer = self.pointer.wrapping_add(written as usize);
+        self.offset += u64::from(written);
+
+        Ok(written as usize)
+    }
+
+    const fn hashtable_hash(&self) -> u64 {
+        // use the original offset as the hash to avoid unnecessary hashing
+        self.original_offset
+    }
+}
+
+impl<'batch, 'ring, I: Iterator<Item = QueueEntry<'batch>>> WriteBatch<'batch, 'ring, I> {
+    fn new(ring: &'ring mut IoUring, fd: RawFd, entries: I) -> Self {
+        let (submitter, sq, cq) = ring.split();
+        Self {
+            fd,
+            submitter,
+            sq,
+            cq,
+            outstanding: self::drop_guard::DropGuard::new(
+                HashTable::with_capacity(COMPLETION_QUEUE_SIZE as usize),
+                hashbrown::HashTable::is_empty,
+                "io-uring write batch terminated with outstanding entries",
+            ),
+            backlog: std::collections::VecDeque::new(),
+            entries,
+            written: 0,
+        }
+    }
+
+    fn is_finished(&self) -> bool {
+        // we never take from the iterator without submitting it to the kernel
+        // or adding it to the backlog, so we don't need to check the iterator
+        self.backlog.is_empty() && self.outstanding.is_empty()
+    }
+
+    fn hashtable_insert(&mut self, entry: QueueEntry<'batch>) {
+        self.outstanding
+            .insert_unique(entry.hashtable_hash(), entry, QueueEntry::hashtable_hash);
```

Comment:
The docs for this method say the behavior is undefined if the value already exists. We only check for the existence of the value in the debug path, so we'd get UB.

I'd rather have this method be fallable and use the `entry` api.

Reply:
This is an unlikely error case and we can check for it, but it really should be checked before we make it into the write batch code.

Reply:
The assumption here is that the node allocator didn't cause a duplicate or overlapping region to be written. But, if we want to be defensive against that then checking for duplicate inserts is also not sufficient. We also need to check for overlapping inserts.

Are we assuming the node allocator is broken here?

Reply:
I'm just wanting a check before we get into an undefined state if somehow this does happen, even in release mode, or alternatively use the entry() api.